builder/actions.py

```diff
@@ -1,63 +1,44 @@
 from collections import defaultdict
 
 from django.db import transaction
+from django.db.models import Count
+from django.utils.text import slugify
 
-from .models import SearchResult
+from codelists import tree_utils
 
 
 def create_codelist(*, owner, name, coding_system_id):
-    return owner.draft_codelists.create(name=name, coding_system_id=coding_system_id)
+    return owner.draft_codelists.create(
+        name=name, slug=slugify(name), coding_system_id=coding_system_id
+    )
 
 
 @transaction.atomic
-def create_search(*, codelist, term, results):
-    search = codelist.searches.create(term=term)
-    search.results.bulk_create(
-        SearchResult(
-            search=search,
-            code=code,
-            matches_term=(code in results["matching_codes"]),
-            is_ancestor=(code in results["ancestor_codes"]),
-        )
-        for code in results["all_codes"]
-    )
+def create_search(*, codelist, term, codes):
+    search = codelist.searches.create(term=term, slug=slugify(term))
+    for code in codes:
+        search.results.create(code=codelist.codes.get_or_create(code=code)[0])
```

Comment:
One for your next pass:

Could the generation of Codelist objects be pulled out of this for loop?

Maybe an inner generator function doing `codelist, _ = …` so you can yield codelist?

It's a slight efficiency improvement because you can then do a bulk_create of Result instances, but I think mainly it would improve readability to show that you're `get_or_create`ing Codelists first.